lib/Modules/FreeRdpHooks/FreeRdpHooks.psm1

```diff
@@ -30,11 +30,19 @@ function Get-Vcredist {
         try {
             Write-JujuWarning "Trying to get vcredist Juju resource"
             $vcredistPath = Get-JujuResource -Resource "vcredist-x64"
-            return $vcredistPath
         } catch {
+            $vcredistPath = $null
             Write-JujuWarning "Failed downloading vcredist resource: $_"
-            Write-JujuWarning "Falling back to file download"
         }
+        if ($vcredistPath) {
+            $bytes = Get-Content $vcredistPath -TotalCount 31 -Encoding Byte
+            $str = [string]::join("", [char[]]$bytes)
+            if ($str -eq $DEFAULT_JUJU_RESOURCE_CONTENT) {
+                Throw "Manually attach the Juju resource 'vcredist-x64' or set the config option 'vcredist-url'."
```

Comment:
So the current workflow is as follows:
- If the user specifies the `installer-url` for the vcredist, than that has priority and it's used even though you have resources or not (this keeps backward compatability with Juju versions < 2.0).
- If the user doesn't specify the installed url for vcredist. Then a try block tries to fetch the resource for vcreds and fails *only* if:
 - the resource was not attached or any other issues as you say (network issue, file corruption, etc);
 - the user has Juju < 2.0 and the `resource-get.exe` doesn't exist
If either of the two failing cases occur, the catch block will print a warning message (and not stop charm execution at all) that the resource get for vcredist was not possible and it falls back to using the default download URL for the vcredist installed (as it was up until adding resources support, backwards compatability again).
So, unless you want to force Juju 2.0 users to use resources and set status such that they must upload the resource for this troublesome vcredist and don't allow them to use the fallback behavior (as for Juju users < 2.0) of a default download URL, then I think this code is fine.